### DISCUSSION TITLE
Way to change or add another version of a package

### DISCUSSION THREAD

**Discussion by RelativeSure** [aidandenlinger/autocopr#7] [Ideas]
Hi,

I can't seem to really understand graphQL and such but I wish for the python script to be possible to add or change to a prerelease version as well. Are you able to help with that? Thanks! 🤗

Also, maybe you can set this repo as a template repo which people can fork from, perhaps?

Best regards and have a good weekend!
RelativeSure

**Comment by aidandenlinger** [comment]
I'm not interested in adding this as a feature as I'm a novice with GraphQL myself and don't have a use for prerelease versions, but I'm happy to give you some direction on what you'd need to do to get prerelease versions!

## Through the REST API

You may want to do this [through the REST API](https://docs.github.com/en/rest?apiVersion=2022-11-28), which is an easier API to work with. It gets more data than you need though, and there's a rate limit if you don't provide a token (I never got around to this, maybe you want to add it?) so it's not perfect, but it's simpler. To use it with the script, *don't* provide a GITHUB_TOKEN:

https://github.com/aidandenlinger/autocopr/blob/09e921f8525fdd9aa5ae18b83862b10b183901f8/.github/workflows/update.yml#L47-L50

Then queries will go through `githubapi/rest.py`. Right now it's [requesting the latest release](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-the-latest-release):

https://github.com/aidandenlinger/autocopr/blob/09e921f8525fdd9aa5ae18b83862b10b183901f8/githubapi/rest.py#L19

I think you can change the url to the query for [getting a list of releases](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#list-releases). Hopefully prereleases are in there and you can work from there. But yeah, you may need to add code to provide a token to the REST API or you may run into rate limiting issues. Right now we only use rest if you don't provide a token:

https://github.com/aidandenlinger/autocopr/blob/09e921f8525fdd9aa5ae18b83862b10b183901f8/autocopr/latestver.py#L13-L22

So if you need to do this, maybe add a `--rest` flag in `autocopr/cli.py`, use that to determine which api to use, and give the rest API access to the token if available. I may get to this feature, but it wouldn't be soon.

## Through GraphQL

Ultimately I think GraphQL is a better solution since it'll only fetch what you need, but it'll definitely take more development time to get the query you need. This is the query that takes a list of ids and gets the releases:

https://github.com/aidandenlinger/autocopr/blob/09e921f8525fdd9aa5ae18b83862b10b183901f8/githubapi/graphql.py#L118-L131

This is the only time I've used GraphQL and I wrote this 2 years ago so I'm a bit rusty, but I know [the Github GraphQL docs](https://docs.github.com/en/graphql) were very useful.

- [`node`](https://docs.github.com/en/graphql/reference/queries#node) is a github query that takes a node id and gets its Node
- I think I used GraphQL's `...` operator to deconstruct the Node if it's a [`Repository`](https://docs.github.com/en/graphql/reference/objects#repository), which all of them were
- I accessed the `latestRelease` field on the [`Repository`](https://docs.github.com/en/graphql/reference/objects#repository) to get a [`Release` object](https://docs.github.com/en/graphql/reference/objects#release)
- and from there I got the `tagName` and `url` from that [`Release` object](https://docs.github.com/en/graphql/reference/objects#release).

I think what you'd want to do is instead of `latestRelease`, you'd want to use the `releases` field, which will give you a [`ReleaseConnection`](https://docs.github.com/en/graphql/reference/objects#releaseconnection) which has a `nodes` field with a `[Release]` (list of releases). I'd assume a prerelease is in there somewhere, hopefully as the first object. I don't know the syntax for this, because these two graphQL queries are the only ones I've written :)

If you're taking this route I'd highly suggest using an editor to test out your queries, I remember going through a lot of trial and error to get my final queries. I used [Altair GraphQL Client](https://altairgraphql.dev/), although that was 2 years ago so I don't know what the best client is now. But something that lets you quickly edit the query and see the results is very very useful!

> Also, maybe you can set this repo as a template repo which people can fork from, perhaps?

I'm not familiar with this, what makes a template repo different from forking the repo?

**Reply by aidandenlinger** [reply]
> So if you need to do this, maybe add a --rest flag in autocopr/cli.py, use that to determine which api to use, and give the rest API access to the token if available. I may get to this feature, but it wouldn't be soon.

I got to this in #8! So if you want to add new features only with the REST API, you can update your github actions to use `--rest` (although you should still pass in a GITHUB_TOKEN to use the API authorized, or else you may face rate limiting)
